File: ui/src/utils/QuadTree.js
import { Rectangle } from "./Rectangle";

export class QuadTree {
  constructor(boundary, capacity) {
    this.boundary = boundary; // Boundary is a rectangle { x, y, w, h }
    this.capacity = capacity; // Max number of objects in a quadrant before subdivision
    this.circles = [];
    this.divided = false;
  }

  subdivide() {
    let x = this.boundary.x;
    let y = this.boundary.y;
    let w = this.boundary.w / 2;
    let h = this.boundary.h / 2;

    let ne = new Rectangle(x + w, y - h, w, h);
    let nw = new Rectangle(x - w, y - h, w, h);
    let se = new Rectangle(x + w, y + h, w, h);
    let sw = new Rectangle(x - w, y + h, w, h);

    this.northeast = new QuadTree(ne, this.capacity);
    this.northwest = new QuadTree(nw, this.capacity);
    this.southeast = new QuadTree(se, this.capacity);
    this.southwest = new QuadTree(sw, this.capacity);

    this.divided = true;
  }

  insert(circle) {
    if (!this.boundary.contains(circle)) {
      return false;
    }

    if (this.circles.length < this.capacity) {
      this.circles.push(circle);
      return true;
    } else {
      if (!this.divided) {
        this.subdivide();
      }

      if (this.northeast.insert(circle)) return true;
      if (this.northwest.insert(circle)) return true;
      if (this.southeast.insert(circle)) return true;
      if (this.southwest.insert(circle)) return true;
    }
  }

  query(range, found) {
    if (!found) {
      found = [];
    }

    if (!this.boundary.intersects(range)) {
      return found;
    } else {
      for (let c of this.circles) {
        if (range.contains(c)) {
          found.push(c);
        }
      }
      if (this.divided) {
        this.northwest.query(range, found);
        this.northeast.query(range, found);
        this.southwest.query(range, found);
        this.southeast.query(range, found);
      }
    }

    return found;
  }
}

File: rust-src/src/quad_tree.rs
use crate::{particle::Particle, rectangle::Rectangle};

#[derive(Clone)]
pub struct QuadTree {
    pub boundary: Rectangle,
    pub capacity: usize,
    pub particles: Vec<Particle>,
    pub is_divided: bool,
    pub northeast: Option<Box<QuadTree>>,
    pub northwest: Option<Box<QuadTree>>,
    pub southeast: Option<Box<QuadTree>>,
    pub southwest: Option<Box<QuadTree>>,
}

impl QuadTree {
    pub fn new(boundary: Rectangle, capacity: usize) -> QuadTree {
        QuadTree {
            boundary,
            capacity,
            particles: Vec::new(),
            is_divided: false,
            northeast: None,
            northwest: None,
            southeast: None,
            southwest: None,
        }
    }

    pub fn subdivide(&mut self) {
        let x = self.boundary.x;
        let y = self.boundary.y;
        let w = self.boundary.w / 2.0;
        let h = self.boundary.h / 2.0;

        let ne = Rectangle::new(x + w, y - h, w, h);
        let nw = Rectangle::new(x - w, y - h, w, h);
        let se = Rectangle::new(x + w, y + h, w, h);
        let sw = Rectangle::new(x - w, y + h, w, h);

        self.northeast = Some(Box::new(QuadTree::new(ne, self.capacity)));
        self.northwest = Some(Box::new(QuadTree::new(nw, self.capacity)));
        self.southeast = Some(Box::new(QuadTree::new(se, self.capacity)));
        self.southwest = Some(Box::new(QuadTree::new(sw, self.capacity)));

        self.is_divided = true;
    }

    pub fn insert(&mut self, particle: &Particle) -> bool {
        if !self.boundary.contains(particle) {
            return false;
        }

        if self.particles.len() < self.capacity {
            self.particles.push(*particle);
            return true;
        } else {
            if !self.is_divided {
                self.subdivide();
            }

            if self.northeast.as_mut().unwrap().insert(particle) {
                return true;
            }
            if self.northwest.as_mut().unwrap().insert(particle) {
                return true;
            }
            if self.southeast.as_mut().unwrap().insert(particle) {
                return true;
            }
            if self.southwest.as_mut().unwrap().insert(particle) {
                return true;
            }
        }

        false
    }

    pub fn query(&mut self, range: &mut Rectangle, found: &mut Vec<Particle>) {
        if !self.boundary.intersects(&range) {
            return;
        }

        for particle in &self.particles {
            if range.contains(&particle) {
                found.push(*particle);
            }
        }

        if self.is_divided {
            self.northeast.as_mut().unwrap().query(range, found);
            self.northwest.as_mut().unwrap().query(range, found);
            self.southeast.as_mut().unwrap().query(range, found);
            self.southwest.as_mut().unwrap().query(range, found);
        }
    }
}
